src/System.Linq/src/System/Linq/Partition.cs

```diff
@@ -389,5 +389,245 @@ public int GetCount(bool onlyIfCheap)
+                Debug.Assert(maxIndexInclusive >= 0);
+                Debug.Assert(minIndexInclusive <= maxIndexInclusive);
+
+                _source = source;
+                _minIndexInclusive = minIndexInclusive;
+                _maxIndexInclusive = maxIndexInclusive;
+            }
+
+            public override Iterator<TSource> Clone()
+            {
+                return new EnumerablePartition<TSource>(_source, _minIndexInclusive, _maxIndexInclusive);
+            }
+
+            public int GetCount(bool onlyIfCheap)
+            {
+                if (onlyIfCheap)
+                {
+                    return -1;
+                }
+
+                int count = _source.Count();
+                if (count <= _minIndexInclusive)
+                {
+                    return 0;
+                }
+
+                return Math.Min(count - 1, _maxIndexInclusive) - _minIndexInclusive + 1;
+            }
+
+            public override bool MoveNext()
+            {
+                // Need to handle -1 and 0 explicitly since we have a default: case.
+                if (_state <= 0)
+                {
+                    Dispose();
+                    return false;
+                }
+
+                switch (_state)
+                {
+                    case 1:
+                        _enumerator = _source.GetEnumerator();
+                        _state = 2;
+                        goto case 2;
+                    case 2:
+                        if (!SkipBeforeFirst(_enumerator))
+                        {
+                            // Reached the end before we finished skipping.
+                            break;
+                        }
+
+                        _state = 3;
+                        goto default;
+                    default:
+                        int taken = _state - 3;
+                        if (taken <= _maxIndexInclusive - _minIndexInclusive && _enumerator.MoveNext())
+                        {
+                            _state++;
+                            _current = _enumerator.Current;
+                            return true;
+                        }
+
+                        break;
+                }
+
+                Dispose();
+                return false;
+            }
+
+            public IPartition<TSource> Skip(int count)
+            {
+                int minIndex = _minIndexInclusive + count;
+                return (uint)minIndex > (uint)_maxIndexInclusive ? EmptyPartition<TSource>.Instance : new EnumerablePartition<TSource>(_source, minIndex, _maxIndexInclusive);
+            }
+
+            public IPartition<TSource> Take(int count)
+            {
+                int maxIndex = _minIndexInclusive + count - 1;
+                return (uint)maxIndex >= (uint)_maxIndexInclusive ? this : new EnumerablePartition<TSource>(_source, _minIndexInclusive, maxIndex);
+            }
+
+            public TSource TryGetElementAt(int index, out bool found)
```

Comment:
I would have avoided optimizing the `TryGet*` functions if I could have, since they take a lot of code and are not as valuable to optimize as the other functions. But, currently they come with implementing `IPartition`.